File: clawsmail-passdecrypt.py
#!/usr/bin/env python
# -*- encoding: utf-8 -*-
#
# Copyright (C) 2013-2016 Colomban Wendling <ban@herbesfolles.org>
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
#
# * Redistributions of source code must retain the above copyright
#   notice, this list of conditions and the following disclaimer.
# * Redistributions in binary form must reproduce the above
#   copyright notice, this list of conditions and the following disclaimer
#   in the documentation and/or other materials provided with the
#   distribution.
# * Neither the name of the author nor the names of its contributors may
#   be used to endorse or promote products derived from this software
#   without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


from Crypto.Cipher import DES
from base64 import standard_b64decode as b64decode
try:
    from ConfigParser import ConfigParser
except ImportError:
    from configparser import ConfigParser


PASSCRYPT_KEY = b'passkey0'


def pass_decrypt(p, key=PASSCRYPT_KEY, mode=DES.MODE_CFB):
    """ Decrypts a password from ClawsMail. """
    if p[0] == '!':  # encrypted password
        buf = b64decode(p[1:])

        """
        If mode is ECB or CBC and the length of the data is wrong, do nothing
        as would the libc algorithms (as they fail early).  Yes, this means the
        password wasn't actually encrypted but only base64-ed.
        """
        if (mode in (DES.MODE_ECB, DES.MODE_CBC)) and ((len(buf) % 8) != 0 or
                                                       len(buf) > 8192):
            d = buf
        else:
            c = DES.new(key, mode=mode, IV=b'\0'*8)
            d = c.decrypt(buf)

        try:
            return d.decode()
        except UnicodeDecodeError:
            return d
    else:  # raw password
        return p


def accountrc_decrypt(filename, key=PASSCRYPT_KEY, mode=DES.MODE_CFB):
    """ Reads passwords from ClawsMail's accountrc file """
    p = ConfigParser()
    p.read(filename)

    for s in p.sections():
        try:
            try:
                address = p.get(s, 'address')
                account = p.get(s, 'account_name')
            except:
                address = '<unknown>'
                account = '<unknown>'

            password = pass_decrypt(p.get(s, 'password'), key, mode=mode)
            print('password for %s, %s is "%s"' % (account, address, password))
        except Exception as e:
            print('Error resolving password for account "%s": %s' % (s, e))


if __name__ == '__main__':
    import os
    from optparse import OptionParser, OptionValueError

    def mode_callback(option, opt, value, parser):
        try:
            parser.values.mode = getattr(DES, 'MODE_'+value.upper())
        except AttributeError:
            raise OptionValueError('Invalid mode "%s"' % value)

    usage = 'Usage: %prog [OPTIONS] ENCRYPTED_PASS1...|FILE...'
    parser = OptionParser(usage=usage)
    parser.add_option('-k', '--key', dest='key', default=PASSCRYPT_KEY,
                      help='Use KEY to decode passwords (8 byte string) '
                           '[%default]',
                      metavar='KEY')
    parser.add_option('-m', '--mode', dest='mode', default=DES.MODE_CFB,
                      type='string', action='callback', callback=mode_callback,
                      help='Use MODE to decrypt DES passwords.  Choose ECB to '
                           'decrypt passwords from FreeBSD installations, CFB '
                           'otherwise [CFB]',
                      metavar='MODE')
    parser.add_option('--freebsd', dest='mode',
                      action='store_const', const=DES.MODE_ECB,
                      help='Use ECB mode as needed for passwords from FreeBSD '
                           '(alias for --mode=ECB)')
    (options, args) = parser.parse_args()

    if len(args) < 1:
        parser.error('missing file or string to decrypt')
    else:
        for a in args:
            if os.path.exists(a):
                accountrc_decrypt(a, key=options.key, mode=options.mode)
            else:
                password = pass_decrypt(a, key=options.key, mode=options.mode)
                print('password "%s" is "%s"' % (a, password))

File: clawsmail-passdecrypt.c
/*
 * Copyright (C) 2012-2013 Colomban Wendling <ban@herbesfolles.org>
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 * * Redistributions of source code must retain the above copyright
 *   notice, this list of conditions and the following disclaimer.
 * * Redistributions in binary form must reproduce the above
 *   copyright notice, this list of conditions and the following disclaimer
 *   in the documentation and/or other materials provided with the
 *   distribution.
 * * Neither the name of the author nor the names of its contributors may
 *   be used to endorse or promote products derived from this software
 *   without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/* build with:
gcc -o clawsmail-passdecrypt clawsmail-passdecrypt.c \
    -I/path/to/clawsmail/src/common \
    `pkg-config --cflags --libs glib-2.0` \
    -lcrypt
*/

#ifndef USE_CLAWSMAIL_SRC
/* we need to build our own version if we want the ECB decryption and not
 * building under FreeBSD -- see clawsmail's sources */
# define USE_CLAWSMAIL_SRC (defined(__FreeBSD__) || !defined(NEED_DES_ECB))
#endif


/* passcrypt from claws-mail -- needs to come before our includes */
#if USE_CLAWSMAIL_SRC
#	include "passcrypt.c"
#endif


#include <string.h>
#include <stdio.h>
#include <glib.h>


#if !USE_CLAWSMAIL_SRC
#	include <rpc/des_crypt.h>
#	define PASSCRYPT_KEY "passkey0"
/* taken from ClawsMail passcrypt.c */
static void passcrypt_decrypt(gchar *password, guint len)
{
	char des_key[8] = PASSCRYPT_KEY;

	des_setparity(des_key);
	ecb_crypt(des_key, password, len, DES_DECRYPT);
}
#endif /* !USE_CLAWSMAIL_SRC */


static gchar *pass_decrypt(const gchar *input)
{
	gchar *output;

	if (input[0] == '!') {
		gsize len;
		gchar *dec;

		dec = (gchar *) g_base64_decode(&input[1], &len);
		passcrypt_decrypt(dec, (guint) len);
		output = dec;
	} else {
		output = g_strdup(input);
	}

	return output;
}

static void accountrc_decrypt(const gchar *filename)
{
	GKeyFile *kf = g_key_file_new();
	GError *error = NULL;

	if (! g_key_file_load_from_file(kf, filename, 0, &error)) {
		fprintf(stderr, "Failed to open file: %s\n", error->message);
		g_error_free(error);
	} else {
		gsize n_groups, i;
		gchar **groups = g_key_file_get_groups(kf, &n_groups);

		for (i = 0; i < n_groups; i++) {
			gchar *input = g_key_file_get_value(kf, groups[i], "password", NULL);
			gchar *address = g_key_file_get_value(kf, groups[i], "address", NULL);
			gchar *account = g_key_file_get_value(kf, groups[i], "account_name", NULL);
			gchar *output = input ? pass_decrypt(input) : NULL;

			printf("password for %s, %s is \"%s\"\n", account, address, output);
			g_free(input);
			g_free(output);
			g_free(address);
			g_free(account);
		}
		g_strfreev(groups);
	}
	g_key_file_free(kf);
}

int main(int argc, char **argv)
{
	int i;

	if (argc < 2) {
		printf("Usage: %s [ENCRYPTED_PASS1...|FILE...]\n", argv[0]);
	}

	for (i = 1; i < argc; i++) {
		if (g_file_test(argv[i], G_FILE_TEST_EXISTS)) {
			/* file exists, let's guess it's a clawsmail accountrc */
			accountrc_decrypt(argv[i]);
		} else {
			gchar *pass = pass_decrypt(argv[i]);

			printf("password \"%s\" is \"%s\"\n", argv[i], pass);
			g_free(pass);
		}
	}

	return 0;
}
